src/source/Response.c

```diff
@@ -625,25 +655,76 @@ SIZE_T postReadCallback(PCHAR pBuffer, SIZE_T size, SIZE_T numItems, PVOID custo
     pCurlApiCallbacks = pCurlRequest->pCurlApiCallbacks;
     uploadHandle = pCurlResponse->pCurlRequest->uploadHandle;
 
-    if (pCurlResponse->paused) {
-        bytesWritten = CURL_READFUNC_PAUSE;
+    if (pCurlResponse->endOfStream || ATOMIC_LOAD_BOOL(&pCurlRequest->requestInfo.terminating)) {
+        DLOGD("Closing connection for upload stream handle: %"
+        PRIu64, uploadHandle);
         CHK(FALSE, retStatus);
     }
 
-    if (pCurlResponse->endOfStream || ATOMIC_LOAD_BOOL(&pCurlRequest->requestInfo.terminating)) {
-        DLOGI("Closing connection for upload stream handle: %" PRIu64, uploadHandle);
-        CHK(FALSE, retStatus);
+    pCurlApiCallbacks->pCallbacksProvider->clientCallbacks.lockMutexFn(
+            pCurlApiCallbacks->pCallbacksProvider->clientCallbacks.customData,
+            pCurlResponse->dataAvailableLock);
+
+    while (bytesWritten == 0 && !ATOMIC_LOAD_BOOL(&pCurlResponse->dataAvailable)) {
+        awaitStatus = CVAR_WAIT(pCurlResponse->dataAvailableCvar, pCurlResponse->dataAvailableLock,
+                                DEFAULT_LOW_SPEED_TIME_LIMIT);
+        if (awaitStatus == STATUS_OPERATION_TIMED_OUT) {
+            DLOGW("Time out...");
+            retStatus = awaitStatus;
+            break;

```

Comment:
Try to structure your logic to avoid break and continue clauses. I always recommend avoiding these keywords